logging: Simplify API for level based logging

ajtowns opened the pull request:
Replace `LogPrint*` functions with severity based logging functions:

 * `LogInfo(...)`, `LogWarning(...)`, `LogError(...)` for unconditional (uncategorised) logging (replaces `LogPrintf`)
 * `LogDebug(CATEGORY, ...)` and `LogTrace(CATEGORY, ...)` for conditional logging (replaces `LogPrint`)
 * `LogPrintLevel(CATEGORY, LEVEL, ...)` for when the level isn't known in advance, or a category needs to be added for an info/warning/error log message (mostly unchanged, but rarely needed)

Logs look roughly as they do now with `LogInfo` not having an `[info]` prefix, and `LogDebug` having a `[cat]` prefix, rather than a `[cat:debug]` prefix. This removes `BCLog::Level::None` entirely -- for `LogFlags::NONE` just use `Level::Info`, for any actual category, use `Level::Debug`.

Adds docs to developer-notes about when to use which level.

Adds `-loglevelalways=1` option so that you get `[net:debug]`, `[all:info]`, `[all:warning]` etc, which might be helpful for automated parsing, or just if you like everything to be consistent. Defaults to off to reduce noise in the default config, and to avoid unnecessary changes on upgrades.

Changes the behaviour of `LogPrintLevel(CATEGORY, BCLog::Level::Info, ...)` to be logged unconditionally, rather than only being an additional optional logging level in addition to trace and debug. Does not change the behaviour of `LogPrintLevel(NONE, Debug, ...)` and `LogPrintLevel(NONE, Trace, ...)` being no-ops.

ajtowns committed:
logging: add LogError, LogWarning, LogInfo, LogDebug, LogTrace

These provide simple and clear ways to write the most common logging
operations:

    LogInfo("msg");
    LogDebug(BCLog::LogFlags::NET, "msg");

    LogError("msg");
    LogWarning("msg");
    LogTrace(BCLog::LogFlags::NET, "msg");

For cases where the level cannot be hardcoded, LogPrintLevel(category,
level, ...) remains available.

ajtowns committed:
logging: Replace uses of LogPrintfCategory

Replace LogPrintfCategory with alternative unconditional log statements.

ajtowns committed:
logging: refactor: pull prefix code out

ajtowns committed:
logging: make [cat:debug] and [info] implicit

ajtowns committed:
logging: Drop BCLog::Level::None

Now that Info-level logging is always logged, there is no further
need for the "None" level, so remove it.

ajtowns commented:
This implements most of my comment https://github.com/bitcoin/bitcoin/pull/25203#issuecomment-1401126785

I figure the main benefit of categories is to be able to selectively turn logging on/off -- but for unconditional logging, you don't have that benefit, so might as well just drop the category entirely, and just have the message be descriptive enough directly. Note that this effectively reverts #25306.

It should be compatible with other in-progress PRs -- it only deprecates the `LogPrint*` functions, it doesn't break them. And it should be easy to finish the conversion with a scripted-diff replacing LogPrintf/LogPrint across the codebase.

This seems like a straightforward way of getting the logging stuff finished up.

cc @jonatack

ajtowns committed:
logging: Log Info messages unconditionally

Previously Info-level logging when a category was specified (via
LogPrintLevel) would only print the corresponding log message if
`-debug=category` were specified, while Info-level logging without a
category would always be printed. Make this more consistent by having
Info messages always be logged, whether they include a category or not.

ajtowns commented:
sample scripted diff code:

```
sed -i 's/LogPrintLevel(\(BCLog::[^,]*\), BCLog::Level::Debug,/LogDebug(\1,/' net_processing.cpp dbwrapper.cpp net.cpp node/txreconciliation.cpp validation.cpp

sed -i 's/LogPrintLevel(\(BCLog::[^,]*\), BCLog::Level::Error, */LogError(/' net.cpp txmempool.cpp

sed -i 's/LogPrintLevel(\(BCLog::[^,]*\), BCLog::Level::Warning, */LogWarning(/' net.cpp

sed -i 's/LogPrintf(/LogInfo(/' net.cpp
sed -i 's/LogPrint(\(BCLog::[^,]*\),/LogDebug(\1,/' net.cpp
```

stickies-v commented:
Concept ACK

maflcko reviewed (approved):
Approach ACK. Makes sense to keep heavily used code snippets short.

This should also fix the issue of Info/Warning/Error messages not having a clear category and forcing the dev to somehow come up with one (or explicitly having to pick `None`)

ajtowns commented:
> Is this ready for review, or any reason why this is still in draft?

Err, perhaps it was because I made some more changes and didn't push them?

ajtowns commented:
> ISTM this conflicts with it, which doesn't help either. It doesn't make sense to conflict with that PR for this one IMO.

The first 6 commits apply cleanly on top of it; though `LogPrintLevel(BCLog::NONE, BCLog::Level::Info, "foo9: %s\n", "bar9"); // no-op` is no-longer a no-op, so the tests need modifying.

jonatack reviewed (commented):
Some thoughts, commit-by-commit.

068de7abc5bda9fd9928d714 The idea behind the severity level logging is to ultimately have only one macro like `Log(category, level)` with one consistent, simple interface for everyone -- both developers updating logging in the code and people who read the logs -- and remove all the other macros when we're ready, simplifying the logging macros, tests, and benchmarks. This commit not only would add several more macros, but also with different interfaces for developers to remember or look up, rather than just one.

https://github.com/bitcoin/bitcoin/pull/28318/commits/94e1b46776e68f26f530c23123e9cecdb661de8a This would move the category from the prefix back into the log message string on an ad hoc basis, would be less consistent, and IMO a step backward.

https://github.com/bitcoin/bitcoin/pull/28318/commits/f20bfc3d95be659cbfa502151dd960c44d315988 Non-urgent approach ACK, I can pull it into the parent in #25203 if you like.

https://github.com/bitcoin/bitcoin/pull/28318/commits/b3bf2fafa678fbde303e7cccff70e3b2a6b66627 This would be a step back, I think, away from consistent explicit logging, to requiring remembering more implicit mental context baggage ("why do I see categories for these logs and not for those?") Also, the prefix allows not needing to put the category manually into the log strings.

https://github.com/bitcoin/bitcoin/pull/28318/commits/53a7f76feb8fee8a00bea245940e46d65892f692 Much of the value from severity level logging will come from making some of the noisy categories (`net` in particular, and maybe validation and mempool) much more useful by selecting which LogPrints become debug vs trace, and in general, by selecting which LogPrintfs become info, warning, or error.

4906f49f0932cc8f6244cff8cfbd79407031e7b2 is an incomplete version of [`118c756` (#25203)](https://github.com/bitcoin/bitcoin/pull/25203/commits/118c7567f62df2b882877590f232242d7c627a05) in the parent pull.

https://github.com/bitcoin/bitcoin/pull/28318/commits/99ac8b174a373d204cd4f2d0d6c1f897d7646526 As discussed in https://github.com/bitcoin/bitcoin/pull/27231#discussion_r1312402046, it seems to make sense to drop the `NONE` category and level enums from `logging.{h,cpp}` when we remove the deprecated macros (as opposed to `0/none` values that are still needed at the init and rpc layers, cf #27231).

https://github.com/bitcoin/bitcoin/pull/28318/commits/cb580599e026c713f7b6abae7a7e873b07b89df8 I agree the config option interface is WIP; Klement Tan who was working on it took a job that diverted him from contributing. It seems incongruent to have `-debug` and `-trace` options but only `-debugexclude`. The plan discussed in one of the step PRs is to combine the functionality of the `-debug` and (help-debug) `-loglevel` options (perhaps to a renamed `-log` config option, which would be more consistent naming-wise with the `logging` RPC), and add log level functionality to that RPC. I haven't done that in the parent PR yet, so it's available to be done!

https://github.com/bitcoin/bitcoin/pull/28318/commits/451f2e76b3293836509c1865222c23ef7b4aacf2 Ok for avoiding a mutex. I have had the nagging thought for a long time that it may not be inconceivable for some users to prefer logging only errors and warnings, thus to avoid coding ourselves into a corner regarding `info` always being logged (which might apply to the preceding commit as well).

ajtowns commented:
> [068de7a](https://github.com/bitcoin/bitcoin/commit/068de7abc5bda9fd9928d7145265ae3a3a851de0) The idea behind the severity level logging is to ultimately have only one macro like `Log(category, level)` with one consistent, simple interface for everyone -- both developers updating logging in the code and people who read the logs

I think there's a higher level goal than that: logs should be convenient and easy for developers and users. Hopefully that's a boring platitude, so I'm going to assume you agree with that?

In particular, I don't think we should prioritise "one function" if it makes reading/writing code tedious and inconvenient, and to me, reading/writing `Log(BCLog::NET, BCLog::Level::Debug, ...)` (with or without the current `PrintLevel`) is pretty ugly. I think it's better to have multiple functions than the other ways I can think of removing the `BCLog::Level::` boilerplate. 

Second, for both users and developers, I think the mix of enabled/disabled categories and global/category-specific levels is super confusing. AFAICS at the highest level we only really want three things:

 * unconditional logging for high-importance information
 * subsystem specific debug logging that can be turned on (in production) to get detailed information about what's going on
 * subsystem specific trace logging to get highly detailed logging that may have severe performance impacts to diagnose specific misbehaviours

> [94e1b46](https://github.com/bitcoin/bitcoin/commit/94e1b46776e68f26f530c23123e9cecdb661de8a) This would move the category from the prefix back into the log message string on an ad hoc basis, would be less consistent, and IMO a step backward.

My reasoning here is that the usefulness of the "category" isn't in telling us what bit of code triggered the log message (we have `-logsourcelocations` for that which is much more precise, and you can generally just grep for the message anyway), but in categorising log messages the user can disable. It's useful for the user to see `[net]` because it tells them they can use `-debug=net` to enable it; it's useful for the user to see `[net:trace]` because it tells them they also need to enable tracing to see that. It's useful to see `[warning]` or `[error]` because those are things they probably want to take action on, and having a consistent tag means they can grep for them more easily. It's not useful to see `[net:info]` because that doesn't tell them anything they can action, it's just noise that makes it harder to read the logs.

So I contend that approach is consistent, it's just consistent over a different (in my opinion more valuable) attribute; namely that categories are solely for conditional logging that the user can disable on a per-subsystem basis.

> [f20bfc3](https://github.com/bitcoin/bitcoin/commit/f20bfc3d95be659cbfa502151dd960c44d315988) Non-urgent approach ACK, I can pull it into the parent in #25203 if you like.

(I don't think any of this is "urgent")

> [b3bf2fa](https://github.com/bitcoin/bitcoin/commit/b3bf2fafa678fbde303e7cccff70e3b2a6b66627) This would be a step back, I think, away from consistent explicit logging, to requiring remembering more implicit mental context baggage ("why do I see categories for these logs and not for those?") Also, the prefix allows not needing to put the category manually into the log strings.

That question is easy to answer: you see categories for log entries that you need to use config options to enable. Likewise in the code: you write `BCLog::HTTP` for `LogDebug` and `LogTrace` entries that will only be shown if the user specifically enables that log category, and you don't write it when that's not the case.

Compare:

1. `LogInfo("Starting HTTP server with %d worker threads\n", rpcThreads);`
2. `LogPrintfCategory(BCLog::HTTP, "Starting server with %d worker threads\n", rpcThreads);`
3. `Log(BCLog::HTTP, BCLog::Level::Info, "Starting server with %d worker threads\n", rpcThreads);`

and the log messages:

1. `Starting HTTP server with 2 worker threads`
2. `[http] Starting server with 2 worker threads`
3. `[http:info] Starting server with 2 worker threads`

For unconditional logging, just writing self-explanatory text is shorter and easier to understand.

For conditional logging, we have to tell the compiler the category, so the `BCLog::HTTP` part isn't really avoidable, but having  functions to cover the two levels of conditional logging at least eliminates the boiler plate.

> [53a7f76](https://github.com/bitcoin/bitcoin/commit/53a7f76feb8fee8a00bea245940e46d65892f692) Much of the value from severity level logging will come from making some of the noisy categories (`net` in particular, and maybe validation and mempool) much more useful by selecting which LogPrints become debug vs trace, and in general, by selecting which LogPrintfs become info, warning, or error.

I think it would make sense for the line between info/warning/error to be:

 * normally, it's a category specific debug log. your log message probably just isn't that important.
 * otherwise, for important but rare messages (eg on startup), use info.
 * for something where the admin probably wants to take action, upgrade that to warning (eg things that trigger `SetMiscWarning`, like your system time being set wrong; or if you've enabled tor but tor connections aren't working)
 * use error for things severe enough that bitcoind will shutdown now (or perhaps just some subsystem, eg a particular wallet that's discovered its db is corrupt/unreadable/unwritable)

That seems pretty easy for both developers and users to keep consistent.

(Actually, maybe it's worth distinguishing "startup" and "recurring-info" messages -- it's useful and okay for startup to be a bit noisy; but info messages should be very rare. Something like `LogInit(...)` for unconditional startup logging, and `LogImportant(...)` for UpdateTip and new outbounds.)

The "if it's usable in production, then do it as `debug`, only use `trace` for stuff that's not usable in production" rule I suggested above would imply that it'd be better to split `net` into multiple categories rather than move it into trace, I think. Concretely, the "send: msg" / "recv: msg" logging in particular could be its own category. `[net:trace]` would only be useful for extra logging that's not suitable for production, maybe like logging every w/txid that's in an INV message, rather than just the first?

> [4906f49](https://github.com/bitcoin/bitcoin/commit/4906f49f0932cc8f6244cff8cfbd79407031e7b2) is an incomplete version of [`118c756` (#25203)](https://github.com/bitcoin/bitcoin/pull/25203/commits/118c7567f62df2b882877590f232242d7c627a05) in the parent pull.

I mean, that PR had been closed for 7 months... That change is also inconsistent with what you recommended in https://github.com/bitcoin/bitcoin/pull/27277#discussion_r1141379531

> [99ac8b1](https://github.com/bitcoin/bitcoin/commit/99ac8b174a373d204cd4f2d0d6c1f897d7646526) As discussed in [#27231 (comment)](https://github.com/bitcoin/bitcoin/pull/27231#discussion_r1312402046), it seems to make sense to drop the `NONE` category and level enums from `logging.{h,cpp}` when we remove the deprecated macros (as opposed to `0/none` values that are still needed at the init and rpc layers, cf #27231).

That commit is about dropping `Level::None` (the alternative that currently gives us "[net] foo" instead of "[net:debug] foo"), not the category `BCLog::NONE` which is what the 27231 discussion is about.

> [cb58059](https://github.com/bitcoin/bitcoin/commit/cb580599e026c713f7b6abae7a7e873b07b89df8) I agree the config option interface is WIP; Klement Tan who was working on it took a job that diverted him from contributing. It seems incongruent to have `-debug` and `-trace` options but only `-debugexclude`. The plan discussed in one of the step PRs is to combine the functionality of the `-debug` and (help-debug) `-loglevel` options (perhaps to a renamed `-log` config option, which would be more consistent naming-wise with the `logging` RPC), and add log level functionality to that RPC. I haven't done that in the parent PR yet, so it's available to be done!

*shrug* That was mostly about giving an api that doesn't expose the global logging level and didn't break tests (which exclude libevent) (and that writing `-debug=net -debug=mempool -log=net:trace -log=mempool:trace` seems unnecessarily clunky). Another approach to consider might be replacing `-debug=all -debugexclude=libevent` with `-debug=all -debug=-libevent`, which would naturally generalise to the same syntax with `-trace`.

I presume the logging RPC would just change to accept `["net"] ["validation"] ["wallet"]` to set net's logging level to debug, validation's to disabled/info, and wallet's to debug+trace, and return `{"net": true, "validation": false, "wallet": "trace"}` or something.

> [451f2e7](https://github.com/bitcoin/bitcoin/commit/451f2e76b3293836509c1865222c23ef7b4aacf2) Concept ACK on avoiding a mutex. I have had the nagging thought for a long time that it may not be inconceivable for some users to prefer logging only errors and warnings, thus to avoid coding ourselves into a corner regarding `info` always being logged (which might apply to the preceding commit as well).

I don't think that's a sensible concern: unconditional logging should be rare enough to not need disabling (eg, connecting to new outbounds, new headers, and new tips for about 400 entries a day), and at that point it's better not to disable them as that gives the user a pretty clear indication their node is still working right. Even for users that only care about errors/warnings, having the UpdateTip context at least is likely useful for giving some clue what state the node was in when it encountered the problem.

ajtowns committed:
logging: add -loglevelalways=1 option

This option tells the logging system to always include a "[cat:level]"
prefix, so [net] becomes [net:debug], LogInfo/LogPrint statements will have
an [all:info] prefix, and LogWarning and LogError logs will become
[all:warning] and [all:error]. This may be easier for automated parsing
of logs, particularly if additional prefixes such as thread or source
location are enabled.

ajtowns committed:
logging: treat BCLog::ALL like BCLog::NONE

ajtowns commented:
Rebased and rearranged.

I've added `-loglevelalways` to give users the choice between low-noise logging / what we have now, where you just have `[net]` or `[warning]` show up, or always having the full category/severity info prefixed `[net:debug]` or `[all:info]` showing up, with the idea that the extra "noise" there might be helpful for automated parsing of logs and similar (and some might just prefer it anyway). In particular, after this PR and with that option set, all log messages specify a category and severity now, so that part of the "severity-based-logging" goal can be considered complete.

I haven't based this on #27231 but it should be easy to rebase either PR once the other has been merged -- there's a conflict in the tests, which have to be edited either way, since 27231 adds test coverage of behaviour that's changed in this PR (namely info logging becoming unconditional).

I've dropped the `-trace` command line argument and the conversion from map to bit flags in an atomic int from this PR; still think stuff like that is worth doing, but it doesn't need to all happen at once.

I've also added docs on when to use the different severities, which seem like a useful thing to have.

I think this should be ready for review/merge at this point.

maflcko reviewed (approved):
Feel free to ignore the nits.

very nice, ACK 757b4c7776f4d11b9860bc93c514d8fea5ef1db 🎺

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: very nice, ACK 757b4c7776f4d11b9860bc93c514d8fea5ef1db 🎺
tV/vqW396UNXRNAWVvgywpEMaq3xHn2/FTivxxEFeBaKfl+dcQWfVKnAiEzqDqsI2B/NlTdjl939rlFW3hC1Bw==
```

</details>

jamesob commented:
Big concept ACK on the interface. This is a nice simplification.

ajtowns commented on src/logging.cpp:
Added a test; tagged as refactor for whatever that's worth

maflcko reviewed (approved):
re-ACK 16dd9042ea81412607c0645e4a3cc152abdc4d9a 😹

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK 16dd9042ea81412607c0645e4a3cc152abdc4d9a 😹
LERMjNl3P8a4k1TfQERdSsQsKXp6qmX2bKUeU8FCwbzPnULd2KQlAdsRWrQjWCs8Ju9RG4vBerN1q+9s/WCpCg==
```

</details>

mzumsande reviewed (commented):
Concept ACK

I like the simplified interface, having to write something like `LogPrintLevel(BCLog::NET, BCLog::Level::Error, ...)` for each log would be annoying, plus I like that there wouldn't be a widely-used function/macro that could log both unconditionally and conditionally depending on parameters, which is a bit of a footgun.

ajtowns commented on doc/developer-notes.md:
```diff
@@ -723,6 +723,44 @@ General Bitcoin Core
+Logging
+-------
+
+The macros `LogInfo`, `LogDebug`, `LogTrace`, `LogWarning` and `LogError` are available for
```
Added notes for these.

ajtowns commented on doc/developer-notes.md:
```diff
@@ -723,6 +723,44 @@ General Bitcoin Core
+  most of the time, and it should be used for log messages that are
+  useful for debugging and can reasonably be enabled on a production
+  system (that has sufficient free storage space). They will be logged
+  if bitcoind is started with `-debug=category` or `-debug=1`.
```
nit notted

stickies-v reviewed (approved):
ACK e60fc7d5d34f23cccbff6e4f5f3d716fa8dad50c

After e60fc7d5d34f23cccbff6e4f5f3d716fa8dad50c, is there a reason not to remove `LogPrintfCategory` completely since it's now only used in tests?

> For the severity level logging, I'd prefer to have a single macro like Log(category, level) with a consistent, simple interface, rather than five macros.

In most other codebases I've worked on, having a per-level logging function was pretty standard. I think the approach suggested here is more developer-friendly than forcing usage of a single `Log` function, especially since the log levels are small in number and not likely to significantly change.

ajtowns commented on src/logging.h:
```diff
@@ -234,22 +237,17 @@ static inline void LogPrintf_(const std::string& logging_function, const std::st
 
 // Log unconditionally.
-#define LogPrintf(...) LogPrintLevel_(BCLog::LogFlags::NONE, BCLog::Level::None, __VA_ARGS__)
+#define LogInfo(...) LogPrintLevel_(BCLog::LogFlags::ALL, BCLog::Level::Info, __VA_ARGS__)
```
Until we can use [std::source_location](https://en.cppreference.com/w/cpp/utility/source_location) (see https://github.com/bitcoin/bitcoin/pull/28999#issuecomment-1852024695) we need to use macros in order to pass `__func__`, `__FILE__` and `__LINE__` information through.

stickies-v replied:
Gha forgout about that. Macro it is, thanks.

jamesob commented on src/logging.h:
```diff
@@ -77,7 +77,6 @@ namespace BCLog {
         Info,      // Default
         Warning,
         Error,
-        None, // Internal use only
```
Minor thing, but there's still a lingering reference in `contrib/devtools/bitcoin-tidy/example_logprintf.cpp`. Not sure if that needs to be updated or not.

jamesob commented on doc/developer-notes.md:
```diff
@@ -723,6 +723,47 @@ General Bitcoin Core
+  Note that `LogPrint(BCLog::CATEGORY, fmt, params...)` is a deprecated
+  alias for `LogDebug`.
+
+- `LogInfo(fmt, params...)` should only be used rarely, eg for startup
```
https://github.com/bitcoin/bitcoin/pull/28318/commits/f7ce5ac08c669ac763e275bb7c82dcfb2b1b6c33

It's somewhat wonky that e.g. `LogInfo()` doesn't take a category, but e.g. `LogDebug` does, and in later commits we actually remove category information from info-equivalent log statements, which seems backwards to me. Ideally, `LogInfo` would _optionally_ take a category. But maybe this can be done later?

stickies-v replied:
> Ideally, LogInfo would optionally take a category.

I had that intuition at first too, but then changed my mind after reading [ajtowns's comment](https://github.com/bitcoin/bitcoin/pull/28318#issuecomment-1702318968) stating:

> My reasoning here is that the usefulness of the "category" isn't in telling us what bit of code triggered the log message (we have -logsourcelocations for that which is much more precise, and you can generally just grep for the message anyway), but in categorising log messages the user can disable. It's useful for the user to see [net] because it tells them they can use -debug=net to enable it; it's useful for the user to see [net:trace] because it tells them they also need to enable tracing to see that. It's useful to see [warning] or [error] because those are things they probably want to take action on, and having a consistent tag means they can grep for them more easily. It's not useful to see [net:info] because that doesn't tell them anything they can action, it's just noise that makes it harder to read the logs.
> 
> So I contend that approach is consistent, it's just consistent over a different (in my opinion more valuable) attribute; namely that categories are solely for conditional logging that the user can disable on a per-subsystem basis.

Given that LogInfo is unconditional, what would be your use case for adding a log category?

jamesob replied:
> Given that LogInfo is unconditional, what would be your use case for adding a log category?

Simply that "info = unconditional" might not always be the case (and is an unusual convention to have hardcoded), and a user could rightly want to see only WARNING+ logs as well as, say, `net` or validation-related INFOs.

stickies-v replied:
I think ajtowns addressed that quite well at the bottom of [this comment](https://github.com/bitcoin/bitcoin/pull/28318#issuecomment-1702318968): 

> I don't think that's a sensible concern: unconditional logging should be rare enough to not need disabling (eg, connecting to new outbounds, new headers, and new tips for about 400 entries a day), and at that point it's better not to disable them as that gives the user a pretty clear indication their node is still working right. Even for users that only care about errors/warnings, having the UpdateTip context at least is likely useful for giving some clue what state the node was in when it encountered the problem.

Seeing warning/error-logs only is easy enough to do with grep, and info logging should be infrequent enough to not take up significant system resources.

jamesob replied:
@stickies-v it's not about system resources; it's about allow the end user to articulate what they see. Not a big deal either way, but I think our current scheme is a little wonky if you compare it to more widely deployed systems e.g. the python logging module or log4j.

stickies-v replied:
> it's about allowing the end user to articulate what they see

which is trivial to do with `grep`? My point about system resources is that theoretically we could always log _everything_ but that would use an unreasonable amount of system resources, hence the nuance.

ryanofsky replied:
> which is trivial to do with `grep`?

This doesn't seem trivial if some log messages have categories and some don't. Like there is no way anymore to only see a list of all the messages only from wallet, or all the messages only from net_processing. I don't think this is terrible, but it does seem weird, and I wonder if any log systems work like this where they go out of their way to define categories, but then choose not to include categories in the most important messages.

ajtowns replied:
> This doesn't seem trivial if some log messages have categories and some don't.

If you want all log messages to have categories for easier grepping (or for importing into a database), there's the `-loglevelalways=1` option introduced in this PR. 

> Like there is no way anymore to only see a list of all the messages only from wallet, 

If you want to focus on messages from some section of the code, then that's what the `-logsourcelocations=1` option is for.

Your wording suggests that you think this is a regression. It's not: we've traditionally used `LogPrintf()` for non-debug messages which also doesn't allow selecting only messages from the wallet, and does not add any consistent tagging.

> I don't think this is terrible, but it does seem weird, and I wonder if any log systems work like this where they go out of their way to define categories, but then choose not to include categories in the most important messages.

I don't think it is actually terribly weird compared to log4j: the equivalent there would be setting up a single global logger, that does info/warn/error levels and always logs all messages, and a bunch of subsystem-specific loggers that do debug/trace levels and that default to off. I think it'd be roughly the same with python: have a global logger for info/warn/error (`logging.warning("boo")`), and child loggers for subsystems for debug/trace (`logging.getLogger('net').debug('did something')`).

I'm not really sure what else to say -- I feel like the arguments stickies-v quoted above were pretty clear, and the result is both simple and useful, even if it seems "weird" or surprising.

jamesob replied:
To clarify, I don't think that anyone is debating that this changeset is an improvement over the status quo. The ask (at least from my end) would be to unify the `Log*` interfaces and allow the optional specification of a category. That way, when migrating existing logging statements over, we're not throwing away category information just because something is INFO. And we support users who want to see WARN only most of the time, but care about, say, `net` INFO messages. That doesn't seem like a crazy thing to ask for.

ryanofsky replied:
> Again, no opinion on this. I think either solution is fine. PRs welcome, I guess?

Thanks, I started playing around with a change that should make it easier to set the category without being verbose, and also start relying less on a global logging instance, which should be useful for kernel applications and tests: 262923c5b37f62a925a2c5611318855009bfb241. Will see how this looks and probably open a PR.

ryanofsky replied:
> Will see how this looks and probably open a PR

Opened #29256 with these changes. Code changes should be complete but there are some test failures so it is a draft.

jamesob reviewed (approved):
ACK e60fc7d5d34f23cccbff6e4f5f3d716fa8dad50c ([`jamesob/ackr/28318.1.ajtowns.logging_simplify_api_for`](https://github.com/jamesob/bitcoin/tree/ackr/28318.1.ajtowns.logging_simplify_api_for))

This is a definite improvement in the new-style logging interface.

I built and reviewed each commit locally, and ran `./test/functional/feature_segwit.py --nocleanup` and reviewed the resulting debug.log to sanity check.

Eventually it would be nice to unify the interface for each logging function (e.g. have the ability to specify category for Info), but personally what is written here would be an improvement over the slightly wordy `LogPrintfCategory()` equivalents that I've never found intuitive.

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e60fc7d5d34f23cccbff6e4f5f3d716fa8dad50c ([`jamesob/ackr/28318.1.ajtowns.logging_simplify_api_for`](https://github.com/jamesob/bitcoin/tree/ackr/28318.1.ajtowns.logging_simplify_api_for))

This is a definite improvement in the new-style logging interface.

I built and reviewed each commit locally, and ran `./test/functional/feature_segwit.py --nocleanup` and reviewed the resulting debug.log to sanity check.

Eventually it would be nice to unify the interface for each logging function (e.g. have the ability to specify category for Info), but personally what is written here would be an improvement over the slightly wordy `LogPrintfCategory()` equivalents that I've never found intuitive.
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmWcfI8ACgkQepNdrbLE
TwWeeBAAiISGHmXYnPZNG+pefl/FmVCdsWNXlPfI1tmV5QNY5e0SwVYXCMCerKYU
n7LQcLJv20rfGpHFclmSqV77eSoZfXTDEY7V+ySERpIWM83H3mPOtmC/KzczfoAd
MklMaQ0GXeyqpZBkU9dstzYR2OvL8/X8Pg8Qsx9+90Cr318AAj+diXePaLyycbkQ
mmbGn36oM69crn6KTilDoDxION1pqC/YAZBLyl9MQ57B4kwML/VM/iLZPUhg9yOf
KbTJ0UVAeSMGrzYeUDlirx0K0KIiy0gEzby6YEPU60YM9YdHIBGIagaWpOIOHAEt
RnWc3VzQryrS3PqHTeTe/PqnHFzc1ukQe1QTGQ7nCJW/RpVijOVjEwgAwPVaB0Zn
9DOZJBWY4WbnCGgfIMwQh35j+qg0Pc7omhmhFb7f6GftSWmEaOytcZef7peEBwyq
9ls7+rBOkkL5CW+8atU+UrRzRjSjvqIUFM14ks5cbgIGfvGzu0uGELaVkyYyuON4
AKZC/kgN+MhYNRINWkZ3d3nVBulc4QqV+ClAMk8WjM8DKz8gstfwSDrCnpQTlBVT
pN/nkGlAHzDeRFi2FypS5ObBo3LHRCzmhqnXbe6IKj+D+8EheFReCXEiuxMNcqpQ
Sp1BmpgFRAe9x/CSUZhicYQThOteZvaDWmL9sz44zewMmk4IV8c=
=Ho6t
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-6.6.8-arch1-1-x86_64-with-glibc2.38

Configured with ./configure 'BDB_LIBS=-L/home/james/src/bitcoin/db4/lib -ldb_cxx-4.8' BDB_CFLAGS=-I/home/james/src/bitcoin/db4/include 'CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare ' --disable-gui-tests --disable-fuzz-binary --enable-wallet --enable-debug --with-daemon --enable-natpmp-default

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare  -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4.1 -msse4.2 -msse4 -msha  i

Compiler version: clang version 16.0.6
```

</p></details>

glozow reviewed (commented):
Concept ACK based on the developer-notes. I'm a fan of having 1 not-too-verbose one that we use most of the time, and then a few others based on severity.

achow101 commented:
ACK e60fc7d5d34f23cccbff6e4f5f3d716fa8dad50c

I also prefer the use of macros/functions for the different categories as it is much less verbose than `LogPrintLevel(CAT, LEVEL, ...)`. This pattern is also similar to what many other logging frameworks do, e.g. python's `logging` module.
